### DISCUSSION TITLE
The Big Plugin System Rework

### DISCUSSION THREAD

**Discussion by Beed-git** [endless-sky/endless-sky#8932] [Balance, Mechanics, and Interactions]
As it stands the plugin system in-game is quite rudimentary. Without using any external tools the options for plugins are limited to installing a plugin manually by adding it into the plugins folder, and enabling and disabling plugins from the main menu. 
I intend to update this system by giving more options to the plugin authors, and providing more feedback and safety to the players. 
Below are a rough collection of my ideas surrounding a rework of how the plugin system works in-game, I would greatly appreciate feedback on what people like and dislike, what should be changed or added, etc. 

- [x] Create a basic plugin metadata file (Done in #8923)
- [ ] Add a dependencies list to the metadata file. This is split into 4 types:
  - [x] Required dependencies: A list of plugins which need to be loaded and enabled for the plugin to work correctly. An example of this would be a sub-plugin for an overhaul plugin.
  - [x] Optional dependencies: A list of plugins which work well with this plugin. For example an extra high-dpi plugin for a plugin which add ships.
  - [x] Conflict dependencies: A list of plugins which are known to conflict and cause issues with this plugin. An example for this would be a Star Wars map overhaul mod would knowingly conflict with a Syndicate space overhaul mod. A plugin author wouldn't be expected to exhaustively fill out this list but this adds the option for them.
  - [ ] Minimum/Maximum game version. This would be a range of known working versions, e.g. 0.10.\*, 0.9.8-0.9.10, 0.\*, etc.
     These dependency changes would be split across 4-5 PRs, one for each dependency type along with potentially a first/final PR to add the fields to the plugin metadata file.
- [ ] Resolve dependencies by sorting plugins through flattening dependency graph, ensure no circular dependencies exist, check all required dependencies exist and are enabled, check no conflicts exist. This would also be split over a few PRs.
- [ ] Defer content loading until after a save file has been set to load. This will allow us to change which plugins are enabled and disabled without loading the plugins and core content.
  - A big question surrounding this is how do we separately load the main menu from the core data. Some options would be to copy the main menu files out of the data folder into a separate coredata folder, or maybe we could manually hardcode the loading for the main menu files.
- [ ] Mount the core endless-sky data as a plugin. This would allow us to have total conversion mods which don't need to override every data file in the data folder, as the player can just disable the endless-sky data. (The plugin author could also add endless-sky to the list of conflict.) This would also solve this issue: #8861
- [ ] Store the list of enabled plugins within the save files. Stops the player from manually having to change which plugins they have enabled, also solves this issue: #661
- [ ] Update the plugin UI to show enabled plugins, disabled plugins, missing plugins, and broken plugins (e.g. circular dependency loop)
- [ ] Move enabled plugins to be per save file rather than global.
- [ ] Add author data to metadata file.
- [ ] Update plugin UI to show author name and copywrite.

Far future updates.
- [ ] Move manifest file information from plugin index into plugin metadata file.
- [ ] Add and remove plugins to the plugin UI as they are added to the plugins folder (so the game doesn't have to be reloaded to show the changes.)
- [ ] Download plugins from plugin index from the plugin UI.

**Comment by Zitchas** [comment]
That is a very comprehensive rework and sounds like it would add a lot of useful functionality.

Right off the bat, the "dependencies" section would provided the needed mechanics for an official "deprecated content" plug-in to be most effective by allowing plug-ins to specify that they require assets found within it. (as mentioned in this PR: https://github.com/endless-sky/endless-sky/pull/8908)

**Comment by shitwolfymakes** [comment]
Concur with Zitchas on this. Ping me on discord if you'd like some help with the implementation

**Comment by DarcyManoel** [comment]
Looks like Github doesn't notify when discussions are mentioned so if you're here, [this](https://github.com/endless-sky/endless-sky/issues/9377) will likely be relevant.
